include/deal.II/matrix_free/fe_point_evaluation.h

```diff
@@ -366,44 +366,86 @@ namespace internal
+       */
+      void
+      compute(const Mapping<dim> &mapping,
+              const typename Triangulation<dim, spacedim>::cell_iterator &cell,
+              const ArrayView<const Point<dim>> &unit_points,
+              const UpdateFlags                  update_flags);
+
+      /**
+       * The points for which the evaluation has been invoked by the last call
+       * to compute().
+       */
+      std::vector<Point<dim>> unit_points;
+
+      /**
+       *  The cell on which the evaluation has been invoked by the last call
+       * to compute().
+       */
+      const typename Triangulation<dim, spacedim>::cell_iterator cell;
+
+      /**
+       * The data computed by this data structure.
+       */
+      internal::FEValuesImplementation::MappingRelatedData<dim, spacedim> data;
+    };
+
+
+
     template <int dim, int spacedim>
     void
-    compute_mapping_data_for_generic_points(
+    MappingData<dim, spacedim>::compute(
       const Mapping<dim> &                                        mapping,
       const typename Triangulation<dim, spacedim>::cell_iterator &cell,
       const ArrayView<const Point<dim>> &                         unit_points,
-      const UpdateFlags                                           update_flags,
-      internal::FEValuesImplementation::MappingRelatedData<dim, spacedim>
-        &mapping_data)
+      const UpdateFlags                                           update_flags)
     {
-      UpdateFlags update_flags_mapping = update_default;
-      // translate update flags
-      if (update_flags & update_jacobians)
-        update_flags_mapping |= update_jacobians;
-      if (update_flags & update_gradients ||
-          update_flags & update_inverse_jacobians)
-        update_flags_mapping |= update_inverse_jacobians;
-      if (update_flags & update_quadrature_points)
-        update_flags_mapping |= update_quadrature_points;
-
-      if (const MappingQ<dim, spacedim> *mapping_q =
-            dynamic_cast<const MappingQ<dim, spacedim> *>(&mapping))
-        {
-          mapping_q->fill_mapping_data_for_generic_points(cell,
-                                                          unit_points,
-                                                          update_flags_mapping,
-                                                          mapping_data);
-        }
-      else if (const MappingCartesian<dim, spacedim> *mapping_cartesian =
-                 dynamic_cast<const MappingCartesian<dim, spacedim> *>(
-                   &mapping))
+      bool needs_update =
+        (cell != this->cell) || unit_points.size() != this->unit_points.size();
+      // expensive check: make sure all points are really the same
+      if (!needs_update)
+        for (unsigned int i = 0; i < unit_points.size(); ++i)
+          if (unit_points[i].distance_square(this->unit_points[i]) > 1e-12)
+            {
+              needs_update = true;
+              break;
+            }
```

Comment:
I guess this check is needed, because all FPE using the same mapping instance can modify it. Wouldn't it be better if there is a master FPE: this is updates the mapping info and returns a non-mutable mapping instance. Also, I am not completely sure about the two reinit functions:
```
reinit(cell, points) -> void
reinit(cell, points, mapping) -> void
```
As far as I understand from the documentation, the `mapping` object in the second reinit function already contains the precomputed mapping. That means we do not need `cell` and `points` in that case because it is already provided by the mapping object, isn't it? Or this mapping object filled (like in `VT::point_values()`) - in that case the documentation is out of date.

Reply:
> Wouldn't it be better if there is a master FPE: this is updates the mapping info and returns a non-mutable mapping instance. 

I agree with @peterrum. It is a nice thought to share the mapping data instead of creating copies (like the code does at the moment), but the shared data should be `const`, otherwise modifying a later FPE might change the behavior of the original one.

> As far as I understand from the documentation, the mapping object in the second reinit function already contains the precomputed mapping.

That is correct. At the moment the second function assumes (and checks) that the mapping data has already been computed, handing over an empty object is not allowed.